### DISCUSSION TITLE
Save all keypresses but continue the trial just with one of two specific keys

### DISCUSSION THREAD

**Discussion by jovitabruening** [jspsych/jsPsych#3133] [Q&A]
Hi all, 

I am quite lost with a specific change I need to make to my experiment. In my jspsych experiment, I must present two stimuli in one trial. However, the participants shall only respond to one of the two tasks. The simultaneous presentation of the two tasks is already implemented (I'm using a table to draw the two stimuli and prompt the participant with an arrow to the according task). Currently I use the type: "html-keyboard-response" for the trials and my saved data also includes information about which task needs to be responded to.
Is there a way to save not only the participant's response to the prompted task but also if he/she presses the button for the not prompted task? However, the trial should only end and the experiment continue once the participant pressed the key for the prompted task. 

Any help or suggestions are highly appreciated :)

**Comment by jodeleeuw** [comment]
Hey @nikbpetrov did you have an extension that recorded all key presses? Or am I remembering the wrong thing?

**Comment by nikbpetrov** [comment]
I do, it's found [here](https://github.com/nikbpetrov/jsPsych-replay), but is a complete overkill for this use case, if you ask me (and it's still draft).

@jovitabruening What I would do here, is preserve the `html-keyboard-response` type trial and set allowed keys to "NO_KEYS". Then, in the `on_load` function, set an event listener to record all key presses as well as their timing. If a specific key is clicked, call `finishTrial` to move on.

Important: Implement RT timing in the same way as is done in the plugin file for `html-keyboard-response`. jsPsych's timing stuff is great so make sure you are implementing it in a consistent way. Also, check the data object at the end of each trial; there are many ways to handle that, e.g. pass data to `finishTrial` or just save all presses/rts in custom data columns.

**Comment by jovitabruening** [comment]
Thank you both for the suggestions and help! 
@nikbpetrov I tried to implement your suggestions and kept the html-keyboard-response and tried to set an event listener in the on_load function. However, I am not sure about the "NO_KEYS" and where to call the finishTrial. 

@both: Maybe I have misunderstood something about how an event listener works. I tried using jsPsych.pluginAPI.getKeyboardResponse. However, what I coded in the on_load function always gets overridden by the basic settings of the trials. In detail: 
- I have two tasks, a task '0' that requires the 'a' and 's' keys to be pressed, and a task '1' that requires the 'k' and 'l' keys to be pressed.
- Accordingly, if the current trial is one of task '0', I want the event listener to catch the first response in case the 'k' or 'l' key was pressed by mistake. But this should not end the trial. The participant should only be able to end the trial by pressing the 'a' or 's' key.
- For task '1' it is the other way round. The event listener should catch when the 'a' or 's' key is pressed, but the trial should only end when the 'k' or 'l' key is pressed. 

In previous experiments, I always provided the trial choices only with the keys of the corresponding tasks (e.g. 's' and 'a' for task '0'). However, jspsych does not seem to enter the event listener and the "valid_responses": [75, 76, 65, 83]" from jsPsych.pluginAPI.getKeyboardResponse does not matter. But if I provide the trial's choices with the keys of both tasks, the trial already terminates when the wrong keys are used and does not wait for the right ones. I cannot figure out how to use NO_KEYS and finishTrial to solve this. How can I either have jspsych enter the event listener or prevent the trial from ending too soon?


Here is my code for the trial. If you have any questions about the code or code related to it, please let me know.

var stim_pres_TSWP = {
		type: "html-keyboard-response", 
		stimulus: jsPsych.timelineVariable('stimulus'),
		choices: jsPsych.timelineVariable('response_keys'),
		data: jsPsych.timelineVariable('data'),

		trial_duration: function () {
			var block_duration = jsPsych.timelineVariable('block_dur', true);
			var block_onset_time = jsPsych.data.get().filter({ unique_tag: 'block_start_time' }).last(1).select('time_elapsed').values;
			var true_block_onset = (block_onset_time[0] + post_onset_gap);
			var first_trial = jsPsych.data.getLastTrialData().filter({ select_data: 'true' }).count();
			if (first_trial == 0) {
				var last_trial_finish_time = true_block_onset;
			} else {
				var last_trial_finish_time = jsPsych.data.get().filter({ select_data: 'true' }).last(1).select('time_elapsed').values;
			}
			var remaining_block_time = block_duration - last_trial_finish_time + true_block_onset;
			return remaining_block_time;
		},

		on_load: function (trial_data) {
			var currentTrialData = jsPsych.data.getLastTrialData().values()[0]; // Get the last trial's data, to determine which task
			currentTask = currentTrialData.task;
			console.log("Current Task:", currentTask);

			if (currentTask === 0) { // if letter task
				var listener_id = jsPsych.pluginAPI.getKeyboardResponse({
					callback_function: function (key_press_data) {
						var continueListening = true; // Flag to control whether to continue listening
						if (continueListening === true) {
							// Check if 'k' or 'l' is pressed
							if (key_press_data.key === 75 || key_press_data.key === 76) {
								jsPsych.data.addDataToLastTrial({
									trial_ends_after_audio: false
								})
								if (key_press_data.key === 75 || key_press_data.key === 76) { // Continue listening
									response1 = NaN;
									response2 = key_press_data.key;
									console.log("Response 1 on_load silent key:", key_press_data.key);
								}
							} else if (key_press_data.key === 65 || key_press_data.key === 83) {
								continueListening = false; // Stop listening if 'a' or 's' is pressed
							}
						}
					},
					valid_responses: [75, 76, 65, 83], // Allow 'k', 'l', 'a', or 's'
					rt_method: 'performance',
					persist: false,
					allow_held_key: false
				});
				console.log("Response 1 on_load:", response1);
				console.log("Response 2 on_load:", response2);

			} else if (currentTask === 1) { // if digit task
				var listener_id = jsPsych.pluginAPI.getKeyboardResponse({
					callback_function: function (key_press_data) {
						var continueListening = true; // Flag to control whether to continue listening
						if (continueListening === true) {
							// Check if 'a' or 's' is pressed
							if (key_press_data.key === 65 || key_press_data.key === 83) {
								jsPsych.data.addDataToLastTrial({
									trial_ends_after_audio: false
								})
								if (key_press_data.key === 65 || key_press_data.key === 83) { // Continue listening
									response1 = key_press_data;
									response2 = NaN;
									console.log("Response 1 on_load silent key:", key_press_data.key);
								}
							} else if (key_press_data.key === 75 || key_press_data.key === 76) {
								continueListening = false; // Stop listening if 'k' or 'l' is pressed
							}
						}
					},
					valid_responses: [65, 83, 75, 76], // Allow 'a', 's', 'k', or 'l'
					rt_method: 'performance',
					persist: false,
					allow_held_key: false
				});
			}
		},

		on_finish: function (trial_data) {
			var lastTrialData = jsPsych.data.getLastTrialData();
			console.log("Response 1 on_finish:", response1);
    		console.log("Response 2 on_finish:", response2);

			var key_press = lastTrialData.select('key_press').values[0];
			var correct_response = lastTrialData.select('correct_response').values[0];
			var correct = key_press == jsPsych.pluginAPI.convertKeyCharacterToKeyCode(correct_response);

			jsPsych.data.addDataToLastTrial({
				correct: correct,
				stimulus1_response: response1,
				stimulus2_response: response2,
				});
			jsPsych.finishTrial();
		}
	}
